### PR TITLE
Add units for utilities and a bug fix

### PR DESCRIPTION
**Pull request by davisshaver**
Adds basic unit test coverage for public functions in the **bolded** files below in the `app/utils` folder.

- bootstrapStore.js
- **buildDeepObject.js**
- chartTypeLoader.js
- **ChartTypeUtils.js**
- **createFormatter.js**
- **errorCodeUtils.js**
- **misc.js**
- **postMessage.js**
- **rawDataHelpers.js**

Also tackles the component ErrorMessage.js, located in `app/components/Header/ErrorMessage/`.

React components are not yet testable due to the bug described in https://github.com/alleyinteractive/simplechart/pull/81.

### CODE REVIEW

File: app/utils/postMessage.js [hunk 1]
```diff
@@ -30,12 +30,12 @@ function _isTopLevelWindow() {
 /**
  * store for message handle callbacks
  */
-const _callbacks = {};
+export const _callbacks = {};
```
Comment (davisshaver) [inline app/utils/postMessage.js]: Export so we can access from test suite

File: app/utils/postMessage.js [hunk 2]
```diff
@@ -30,12 +30,12 @@ function _isTopLevelWindow() {
...
 
 /**
  * Setup postMessage receive callbacks
  */
-export function setupPostMessage() {
+export function setupPostMessage(window = window) {
```
Comment (davisshaver) [inline app/utils/postMessage.js]: Adding `window` as optional parameter helps with testability by allowing us to pass in simulated window conditions. (Inspired by [this stack thread](http://stackoverflow.com/a/8922356).)

File: app/utils/misc.js [hunk 3]
```diff
@@ -33,7 +33,7 @@ export function debounce(func, ...theArgs) {
  * @return bool True if multi-series, false if single-series
  */
 export function dataIsMultiSeries(data) {
-  return ('undefined' !== typeof data[0].key &&
+  return ('undefined' !== typeof data[0].keys &&
```
Comment (davisshaver) [inline app/utils/misc.js]: 🐛 fix?

File: tests/utils/postMessage.sendMessage.unit.test.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+/* global describe test expect jest */
+/* eslint no-undef: "error" */
+/* eslint-disable no-console */
+
+const postMessage = require('../../app/utils/postMessage.js');
```
Comment (joshkadis) [inline tests/utils/postMessage.sendMessage.unit.test.js]: here you could do 
```
import { sendMessage } from '../../app/utils/postMessage.js';
...
sendMessage('test', 'data');
```

File: tests/utils/postMessage.setupPostMessage.unit.test.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/* global describe test expect jest */
+/* eslint no-undef: "error" */
+
+const postMessage = require('../../app/utils/postMessage.js');
```
Comment (joshkadis) [inline tests/utils/postMessage.setupPostMessage.unit.test.js]: same as with `postMessage.sendMessage`

File: tests/utils/rawDataHelpers.unit.test.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    expect(parsed[1]).toEqual(fields);
+    expect(parsed[2]).toEqual([]);
+  });
+  test('papa single', () => {
+    const parsedSingle = rawDataHelpers.parseRawData(Papa, rawDataSingle);
+    expect(parsedSingle[0]).toEqual(parsedDataSingle);
+    expect(parsedSingle[1]).toEqual(fieldsSingle);
+    expect(parsedSingle[2]).toEqual([]);
+  });
+  test('papa bad', () => {
```
Comment (joshkadis) [inline tests/utils/rawDataHelpers.unit.test.js]: I get this test _all the time_.

Comment (davisshaver) [inline tests/utils/rawDataHelpers.unit.test.js]: papa-john-pepp_0

**Comment by joshkadis** [comment]
:horse: to @davisshaver 's stuff

**Comment by davisshaver** [comment]
@joshkadis Ok, fixed a few things.

First that CSS issue. https://github.com/alleyinteractive/simplechart/pull/80/commits/e66ee6ad1096aed1a5f207b5662954eef9a1ba16

Also took me a bit of time but I identified a weird bug with the jest tests. For our error message tests, separate the shallow from deep tests was necessary to get the shallow to work correctly.

In the deep test I also found that I had to have something within the `<ErrorMessage>` element, I used `{false}`, but otherwise I was getting an error about empty child.

```
Warning: Failed prop type: The prop `children` is marked as required in `ErrorMessage`, but its value is `undefined`.
```

Seems to be this line: https://github.com/alleyinteractive/simplechart/blob/master/app/components/Header/ErrorMessage/index.js#L80

Should we set a child if there's a prop.code passed? Or drop as an element required prop?

**Comment by davisshaver** [comment]
Oh I see what you were trying to do with `e000` now. Would it makes sense to drop `children` as required for the element?

**Comment by davisshaver** [comment]
`npm run test-lint`, `npm run test`, and `run-p start test-integration -r --silent` should all pass now.

**Comment by davisshaver** [comment]
@joshkadis This this is good to merge. Wanna take another look? Once we have it in, we can add Travis and expand for other components.

**Comment by davisshaver** [comment]
@joshkadis Want to take a 👀  before merge?

**Comment by joshkadis** [comment]
@davisshaver yeah, let's revisit this one. thanks for putting it back on my radar.

**Pull request closed without merging**